msm/src/circuit_design/witness.rs

```diff
@@ -10,14 +10,24 @@ use crate::{
 use ark_ff::PrimeField;
 use kimchi::circuits::domains::EvaluationDomains;
 use log::debug;
-use std::{collections::BTreeMap, iter};
+use std::{collections::BTreeMap, iter, marker::PhantomData};
 
-/// Witness builder environment. Operates
-pub struct WitnessBuilderEnv<F: PrimeField, const CIX_COL_N: usize, LT: LookupTableID> {
+/// Witness builder environment. Operates on multiple rows at the same
+/// time. `CIx::N_COL` must be equal to `N_COL`; passing these two
+/// separately is due to a rust limitation.
+pub struct WitnessBuilderEnv<
+    F: PrimeField,
+    CIx: ColumnIndexer,
+    const N_COL: usize,
+    const N_REL: usize,
+    const N_DSEL: usize,
+    const N_FSEL: usize,
+    LT: LookupTableID,
+> {
     /// The witness columns that the environment is working with.
     /// Every element of the vector is a row, and the builder is
     /// always processing the last row.
-    pub witness: Vec<Witness<CIX_COL_N, F>>,
+    pub witness: Vec<Witness<N_COL, F>>,
```

Comment:
Like, here maybe you could say instead `Witness<CIx::N_COL,F>` ? Not sure as I haven't tried with constant generics, but maybe worth trying.

Reply:
this does not work at least in stable... which is a very useful feature to have tbh. So, unfortunately.
```
error: generic parameters may not be used in const operations
  --> msm/src/circuit_design/witness.rs:30:32
   |
30 |     pub witness: Vec<Witness<{ CIx::N_COL }, F>>,
   |                                ^^^^^^^^^^ cannot perform const operation using `CIx`
   |
   = note: type parameters may not be used in const expressions
```

Reply:
It will work on `nightly` though, which makes me super envious :( :( :( 

https://stackoverflow.com/questions/65461672/rust-constant-expression-depends-on-a-generic-parameter

Reply:
Mmmm I see, well let's wait until they allow this in stable 👍🏻